File: BFS/616.Course Schedule II/Solution.py
from collections import deque, defaultdict

class Solution:
    """
    @param: numCourses: a total of n courses
    @param: prerequisites: a list of prerequisite pairs
    @return: the course order
    """
    def findOrder(self, numCourses, prerequisites):
        # write your code here
        courseIndegree = self._getCourseIndegree(prerequisites)
        neighbors = self._getNeighbors(prerequisites)

        start_courses = [course for course in range(numCourses) if courseIndegree[course] == 0]
        queue = deque(start_courses)
        order = []

        while queue:
            course = queue.popleft()
            order.append(course)

            for neighbor in neighbors[course]:
                courseIndegree[neighbor] -= 1
                if courseIndegree[neighbor] == 0:
                    queue.append(neighbor)

        return order if len(order) == numCourses else []


    def _getCourseIndegree(self, prerequisites):
        courseIndegree = defaultdict(int)

        for i, j in prerequisites:
            courseIndegree[i] += 1

        return courseIndegree

    def _getNeighbors(self, prerequisites):
        neighbors = defaultdict(list)

        for i, j in prerequisites:
            neighbors[j].append(i)

        return neighbors

File: BFS/616.Course Schedule II/Solution.java
public class Solution {
    /*
     * @param numCourses: a total of n courses
     * @param prerequisites: a list of prerequisite pairs
     * @return: the course order
     */
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        // write your code here

        int[] result = new int[numCourses];
        int[] indegree = new int[numCourses];
        List[] order = new ArrayList[numCourses];  // Array of ArrayList
        int added = 0;

        // Get prerequisites
        for (int i = 0; i < numCourses; i++) {
            order[i] = new ArrayList<Integer>();
        }
        
        // Update indegree of all courses
        for (int[] pair : prerequisites) {
            indegree[pair[0]]++;
            order[pair[1]].add(pair[0]);
        }
        
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                queue.offer(i);
            }
        }

        // BFS
        while (!queue.isEmpty()) {
            
            int prerequisite = queue.poll();
            result[added] = prerequisite;

            
            for (Object c : order[prerequisite]) {
                int course = (int) c;

                indegree[course]--;

                if (indegree[course] == 0) {
                    queue.offer(course);
                }
    
            }
            added++;
        }

        if (added == numCourses) {
            return result;
        } else {
            return new int[0];
        }
    }

}